### PR TITLE
Add Controller.open_dir_entry function to open a dir_entry

### PR DESCRIPTION
**Pull request by ceigel**
This fixes issue #18
When calling iter_dir the DirEntry can be extracted via the func lambda. There is no function opening a file from DirEntry. This PR adds support for that.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -796,7 +796,24 @@ where
 //
 // ****************************************************************************
 
-// None
+/// Transform mode variants (ReadWriteCreate_Or_Append) to simple modes ReadWriteAppend
```
Comment (thejpster) [inline src/lib.rs:799]: Can you explain the reason behind this logic?

File: src/filesystem.rs [hunk 2]
```diff
@@ -268,6 +270,17 @@ impl ShortFileName {
     const FILENAME_BASE_MAX_LEN: usize = 8;
     const FILENAME_MAX_LEN: usize = 11;
 
+    /// Get base name (name without extension) of file name
+    pub fn base_name(&self) -> Result<&str, FilenameError> {
```
[the remark this answers is not in the document]
Comment (ceigel) [reply src/filesystem.rs:274]: Ok. I didn't consider that. But how would this function here change if there would be a different encoding?

File: src/lib.rs (hunk 1 above)
Comment (ceigel) [reply src/lib.rs:799]: I'm confused by your question. This is just the lines 487-500 from the old file extracted in a separate function. What is missing?

Comment (thejpster) [reply src/lib.rs:799]: Sorry, I was just reading this function and I couldn't work out what it was trying to do or why!

File: src/filesystem.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (thejpster) [reply src/filesystem.rs:274]: Well either we return raw bytes, or we find out the encoding , convert to Unicode and return a String.

[the remark this answers is not in the document]
Comment (thejpster) [reply src/filesystem.rs:274]: I also recall that the filenames are space padded, which might be unexpected.

**Comment by ceigel** [comment]
@thejpster I addressed (think at least:) ) your comments. Could you please review again?

File: src/filesystem.rs [hunk 3]
```diff
@@ -268,6 +270,22 @@ impl ShortFileName {
     const FILENAME_BASE_MAX_LEN: usize = 8;
     const FILENAME_MAX_LEN: usize = 11;
 
+    /// Get base name (name without extension) of file name
+    pub fn base_name(&self) -> &[u8] {
+        let mut first_space: usize = 0;
```
Comment (thejpster) [inline src/filesystem.rs:275]: Default to 8 here, as a file name may not have any spaces. Worth adding some unit tests for this to catch the edge cases!

File: src/filesystem.rs [hunk 4]
```diff
@@ -268,6 +270,22 @@ impl ShortFileName {
...
+            if self.contents[idx] == b' ' {
+                break;
+            }
+        }
+        &self.contents[..first_space]
+    }
+
+    /// Get base name (name without extension) of file name
+    pub fn extension(&self) -> &[u8] {
+        &self.contents[Self::FILENAME_BASE_MAX_LEN..]
```
Comment (thejpster) [inline src/filesystem.rs:287]: Yeah, extensions can be space padded to.

File: src/filesystem.rs (hunk 3 above)
Comment (thejpster) [reply src/filesystem.rs:275]: *scroll*

Lol, there they are ;)

File: src/filesystem.rs [hunk 5]
```diff
@@ -735,6 +753,20 @@ mod test {
         assert_eq!(sfn, ShortFileName::create_from_str("HELLO.TXT").unwrap());
     }
 
+    #[test]
+    fn filename_get_extension() {
+        let mut sfn = ShortFileName::create_from_str("hello.txt").unwrap();
+        assert_eq!(sfn.extension().unwrap(), "TXT");
+        sfn = ShortFileName::create_from_str("hello").unwrap();
+        assert_eq!(sfn.extension().unwrap(), "   ");
```
Comment (thejpster) [inline src/filesystem.rs:761]: No extension should give the empty string.

Test cases:

a.x
abc.xy
abcdefgh.x
abcdefgh.xyz

File: src/lib.rs [hunk 6]
```diff
@@ -796,7 +796,25 @@ where
 //
 // ****************************************************************************
 
-// None
+/// Transform mode variants (ReadWriteCreate_Or_Append) to simple modes ReadWriteAppend or
+/// ReadWriteCreate
+fn solve_mode_variant(mode: Mode, dir_entry_is_some: bool) -> Mode {
```
Comment (thejpster) [inline src/lib.rs:801]: Could we add some test cases for this?

File: src/lib.rs [hunk 7]
```diff
@@ -436,86 +436,38 @@ where
         }
     }
 
-    /// Open a file with the given full path. A file can only be opened once.
-    pub fn open_file_in_dir(
+    /// Open a file from DirEntry. This is obtained by calling iterate_dir. A file can only be opened once.
+    pub fn open_dir_entry(
```
Comment (thejpster) [inline src/lib.rs:440]: Can we add a test for this?

**Review by thejpster — DISMISSED** [review]

**Comment by thejpster** [comment]
Er, this says "Expected — Waiting for status to be reported", but I don't know what Github is waiting for. Any ideas?

**Comment by ceigel** [comment]
No idea. Should I rebase? How do we ask it to recheck?

**Comment by thejpster** [comment]
I think the problem is the GitHub Action only runs on push, not PRs, but the PR is gated on a passing build.

**Comment by thejpster** [comment]
paging @eldruin or @jamesmunns as I don't know what I'm doing any more.

**Comment by eldruin** [comment]
bors ist not enabled for this repository. Or am I missing something?

**Comment by eldruin** [comment]
@thejpster I have no permissions for this repo. Can you grant me them?

**Review by eldruin — APPROVED** [review]

**Comment by eldruin** [comment]
@ceigel CI was not active for pull requests. Could you rebase this to mater?

**Comment by ceigel** [comment]
@eldruin my pr was already based on the latest master. The last commit in master I see is 5d7a574. I see you have a new merge on develop. The last commit on develop is ffe422b. Should I base this PR on develop or master?

**Comment by eldruin** [comment]
Oops, sorry. develop seems to be the default branch here so I would say we should add the stuff there and eventually merge to master when doing a new release.

**Review by eldruin — APPROVED** [review]

**Comment by eldruin** [comment]
Thank you @ceigel!
